In [2]:
String PROBLEM_DATA = "./src/problem300.txt";

public static Set<Integer> getProblems(File file) {
    Set<Integer> problems = new HashSet<>();
    try {
        Scanner scan = new Scanner(file);
        while (scan.hasNext()) {
            int i = scan.nextInt();
            problems.add(i);
        }
    } catch (Exception ex) {
        ex.printStackTrace();
    }
    return problems;
}

public void addProblem(int problem){
    try {
        File currentProblems = new File(PROBLEM_DATA);
        Set<Integer> problems = getProblems(currentProblems);
        problems.add(problem);
        currentProblems.createNewFile();
        FileWriter writer = new FileWriter(currentProblems);
        for(int i : problems) {
            writer.write(String.valueOf(i) + "\n");
        }
        writer.flush();
        writer.close();
    }catch (Exception ex) {
        ex.printStackTrace();
    }
}

public void addProblems(int[] problem){
    try {
        File currentProblems = new File(PROBLEM_DATA);
        Set<Integer> problems = getProblems(currentProblems);
        for(int i : problem) {
            problems.add(i);
        }
        currentProblems.createNewFile();
        FileWriter writer = new FileWriter(currentProblems);
        for(int i : problems) {
            writer.write(String.valueOf(i) + "\n");
        }
        writer.flush();
        writer.close();
    }catch (Exception ex) {
        ex.printStackTrace();
    }
}

public void removeProblem(int problem){
    try {
        File currentProblems = new File(PROBLEM_DATA);
        Set<Integer> problems = getProblems(currentProblems);
        problems.remove(problem);
        currentProblems.createNewFile();
        FileWriter writer = new FileWriter(currentProblems);
        for(int i : problems) {
            writer.write(String.valueOf(i) + "\n");
        }
        writer.flush();
        writer.close();
    }catch (Exception ex) {
        ex.printStackTrace();
    }
}

public void removeProblems(int[] problem){
    try {
        File currentProblems = new File(PROBLEM_DATA);
        Set<Integer> problems = getProblems(currentProblems);
        for(int i : problem) {
            problems.remove(i);
        }
        currentProblems.createNewFile();
        FileWriter writer = new FileWriter(currentProblems);
        for(int i : problems) {
            writer.write(String.valueOf(i) + "\n");
        }
        writer.flush();
        writer.close();
    }catch (Exception ex) {
        ex.printStackTrace();
    }
}

public void printMissing(int start, int end){
    if(start > end){
        return;
    }
    start = Math.max(0, start);
    int total = end - start + 1;

    File currentProblems = new File(PROBLEM_DATA);
    Set<Integer> problems = getProblems(currentProblems);
    ArrayList<Integer> remain = new ArrayList<>();
    for(int i = start; i <= end; i++ ) {
        if( problems.contains(i) ) {
            continue;
        }
        remain.add(i);
    }
    String line1 = "Total " + String.valueOf(total) + " from " + String.valueOf(start) +" to " + String.valueOf(end);
    String line2 = " Left " + String.valueOf(remain.size());
    System.out.println(line1 + line2);
    System.out.println("Printing Missing Problems");
    int count = 0;
    for(int i : remain) {
        System.out.print(String.valueOf(i) + " ");
        count++;
        if(count > 10) {
            count = 0;
            System.out.println();
        }
    }
}

public boolean isInTheFile(int number) {
    File currentProblems = new File(PROBLEM_DATA);
    Set<Integer> problems = getProblems(currentProblems);
    if(problems.contains(number)){
        System.out.println(String.valueOf(number) + " already IN.");
        return true;
    }else{
        System.out.println(String.valueOf(number) + " NOT IN.");
        return false;
    }
}


In [3]:
String PROBLEMS_UNION = "./src/problemsUnion.txt";

// union find helpers 
public void addNode(Map<Integer, Integer> roots, int node) {
    if (roots.containsKey( node )) {
        return;
    }
    roots.put( node,  node);
}

public int findRoot( Map<Integer, Integer> roots, int node) {
    int parentNode = roots.getOrDefault(node, node);
    if (node == parentNode) {
        return parentNode;
    }
    int rootNode = findRoot(roots, parentNode);
    roots.put(node, rootNode);
    return roots.getOrDefault(node, node);
}

public void unionFind(Map<Integer, Integer> roots, int node1, int node2) {
    addNode( roots, node1); 
    addNode( roots, node2);
    int root1 = findRoot(roots, node1);
    int root2 = findRoot(roots, node2);
    if(root1 < root2) {
        roots.put(root2, root1);
    }else{
        roots.put(root1, root2);
    }

}

// union find from file
public static Map<Integer, Integer>  getUnion(File file) {
    Map<Integer, Integer> roots = new HashMap<>();
    try {
        Scanner scan = new Scanner(file);
        while (scan.hasNext()) {
            int p1 = scan.nextInt();
            int p2 = scan.nextInt();
            unionFind(roots, p1, p2);
        }
    } catch (Exception ex) {
        ex.printStackTrace();
    }
    return roots;
}

// helpers of add and remove
public TreeMap<Integer, TreeSet<Integer>> rootsToUnion(Map<Integer, Integer> roots) {
    TreeMap<Integer, TreeSet<Integer>> union = new TreeMap<>();

    for(int i : roots.keySet()){
        int root = roots.get(i);
        if( !union.containsKey(root) ) {
            union.put(root, new TreeSet<Integer>());
            union.get(root).add(root);
        }
        union.get(root).add(i);
    }
    return union;
}

public ArrayList< ArrayList<Integer> > unionToGroup(TreeMap<Integer, TreeSet<Integer>> union) {
    ArrayList< ArrayList<Integer> > group = new ArrayList<>();
    for(TreeSet<Integer> set : union.values()) {
        group.add(new ArrayList<Integer>(set));
    }
    return group;
}

public void saveUnion(ArrayList< ArrayList<Integer> > union) {
    try {
        File curUnion = new File(PROBLEMS_UNION);
        curUnion.createNewFile();
        FileWriter writer = new FileWriter(curUnion);
        for(ArrayList<Integer> list : union) {
            int root = list.get(0);
            for(int i = 0; i < list.size(); i++ ) {
                writer.write(String.valueOf(root) + " "+ String.valueOf( list.get(i) ) +"\n");
            }
        }
        writer.flush();
        writer.close();
    }catch (Exception ex) {
        ex.printStackTrace();
    }
}

// add
public void addToUnion(int[] numbers) {
    addProblems(numbers);
    Set<Integer> tempSet = new TreeSet<>();
    for(int i : numbers){
        tempSet.add(i);
    }
    
    ArrayList<Integer> tempList = new ArrayList<>(tempSet);
    File problemsUnionFile = new File(PROBLEMS_UNION);
    Map<Integer, Integer> roots = getUnion(problemsUnionFile);
    for(int i : tempList) {
        unionFind(roots, tempList.get(0), i);
    }

    TreeMap<Integer, TreeSet<Integer>> union =  rootsToUnion(roots);
    ArrayList< ArrayList<Integer> > group = unionToGroup(union);
    saveUnion(group);
    System.out.println("添加成功！");
}

public void addUnionPair(int a, int b) {
    addToUnion(new int[]{a,b});
}

// remove
public void removeFromUnion(int[] numbers) {
    removeProblems(numbers);
    Set<Integer> tempSet = new TreeSet<>();
    for(int i : numbers){
        tempSet.add(i);
    }

    File problemsUnionFile = new File(PROBLEMS_UNION);
    Map<Integer, Integer> roots = getUnion(problemsUnionFile);
    
    for(int i : tempSet) {
        if( !roots.containsKey(i)){
            continue;
        }
        if( roots.get(i) == i ) { // i is a root;
            Set<Integer> leftElements = new HashSet<>();
            int randomJ = 0;
            for(int j : roots.keySet() ) {
                if( j == i ) continue;
                randomJ = j;
                if( roots.get(j) == i ) {
                    leftElements.add(j);
                }
            }
            for(int j : leftElements) {
                roots.remove(j);
            }
            for(int j : leftElements){
                unionFind(roots, j, randomJ);
            }
        }
        roots.remove(i);
    }

    TreeMap<Integer, TreeSet<Integer>> union =  rootsToUnion(roots);
    ArrayList< ArrayList<Integer> > group = unionToGroup(union);
    saveUnion(group);
}

public void printUnion(){
    File problemsUnionFile = new File(PROBLEMS_UNION);
    Map<Integer, Integer> roots = getUnion(problemsUnionFile);
    TreeMap<Integer, TreeSet<Integer>> union =  rootsToUnion(roots);
    ArrayList< ArrayList<Integer> > group = unionToGroup(union);
    int numberOfGroups = group.size();
    int numberOfProblems = 0;
    
    System.out.println("Printing Problem Unions: ");
    int count = 1;
    for(ArrayList<Integer> list : group ){
        System.out.println("Group " + String.valueOf(count++));
        numberOfProblems += list.size();
        StringBuilder curGroup = new StringBuilder();
        curGroup.append("      ");
        for(int i : list) {
            curGroup.append(String.valueOf(i) + " ");
        }
        System.out.println(curGroup.toString());
    }
    System.out.println("Total Groups: " + String.valueOf(numberOfGroups));
    System.out.println("Total Problems: " + String.valueOf(numberOfProblems));
}

public void findUnion(int number){
    File problemsUnionFile = new File(PROBLEMS_UNION);
    Map<Integer, Integer> roots = getUnion(problemsUnionFile);
    if(!roots.containsKey(number)){
        return;
    }
    TreeMap<Integer, TreeSet<Integer>> union =  rootsToUnion(roots);
    TreeSet<Integer> set = union.get( roots.get(number) );
    
    System.out.println("Printing the set that problem "+String.valueOf(number) + " in: ");
    StringBuilder curSet = new StringBuilder();
    for(int i : set) {
        curSet.append(String.valueOf(i) + " ");
    }
    System.out.println(curSet.toString());
    System.out.println("Total "+ String.valueOf(set.size())+" problems in this set.");
}

In [4]:
//用例
addToUnion(new int[]{});
removeFromUnion(new int[]{});
printUnion();
findUnion(1);

java.io.FileNotFoundException: ./src/problem300.txt (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:216)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at java.base/java.util.Scanner.<init>(Scanner.java:639)
	at REPL.$JShell$32.getProblems($JShell$32.java:20)
	at REPL.$JShell$34.addProblems($JShell$34.java:20)
	at REPL.$JShell$47.addToUnion($JShell$47.java:19)
	at REPL.$JShell$52.do_it$($JShell$52.java:16)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.github.spencerpark.ijava.execution.IJavaExecutionControl.lambda$execute$1(IJavaExecutionControl.jav

添加成功！


java.io.FileNotFoundException: ./src/problem300.txt (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:216)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at java.base/java.util.Scanner.<init>(Scanner.java:639)
	at REPL.$JShell$32.getProblems($JShell$32.java:20)
	at REPL.$JShell$36.removeProblems($JShell$36.java:20)
	at REPL.$JShell$49.removeFromUnion($JShell$49.java:19)
	at REPL.$JShell$53.do_it$($JShell$53.java:16)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.github.spencerpark.ijava.execution.IJavaExecutionControl.lambda$execute$1(IJavaExecutionCon

Printing Problem Unions: 
Total Groups: 0
Total Problems: 0


java.io.FileNotFoundException: ./src/problemsUnion.txt (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:216)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at java.base/java.util.Scanner.<init>(Scanner.java:639)
	at REPL.$JShell$43.getUnion($JShell$43.java:21)
	at REPL.$JShell$51.findUnion($JShell$51.java:19)
	at REPL.$JShell$55.do_it$($JShell$55.java:16)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.github.spencerpark.ijava.execution.IJavaExecutionControl.lambda$execute$1(IJavaExecutionControl.java:95)
	at java.base/java.util.concurrent.FutureTask.r

In [5]:
// From 1 - 200
printMissing(1, 200);

java.io.FileNotFoundException: ./src/problem300.txt (No such file or directory)
	at java.base/java.io.FileInputStream.open0(Native Method)
	at java.base/java.io.FileInputStream.open(FileInputStream.java:216)
	at java.base/java.io.FileInputStream.<init>(FileInputStream.java:157)
	at java.base/java.util.Scanner.<init>(Scanner.java:639)
	at REPL.$JShell$32.getProblems($JShell$32.java:20)
	at REPL.$JShell$37.printMissing($JShell$37.java:25)
	at REPL.$JShell$56.do_it$($JShell$56.java:16)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.github.spencerpark.ijava.execution.IJavaExecutionControl.lambda$execute$1(IJavaExecutionControl.java:95)
	at java.base/java.util.concurrent.FutureTas

Total 200 from 1 to 200 Left 200
Printing Missing Problems
1 2 3 4 5 6 7 8 9 10 11 
12 13 14 15 16 17 18 19 20 21 22 
23 24 25 26 27 28 29 30 31 32 33 
34 35 36 37 38 39 40 41 42 43 44 
45 46 47 48 49 50 51 52 53 54 55 
56 57 58 59 60 61 62 63 64 65 66 
67 68 69 70 71 72 73 74 75 76 77 
78 79 80 81 82 83 84 85 86 87 88 
89 90 91 92 93 94 95 96 97 98 99 
100 101 102 103 104 105 106 107 108 109 110 
111 112 113 114 115 116 117 118 119 120 121 
122 123 124 125 126 127 128 129 130 131 132 
133 134 135 136 137 138 139 140 141 142 143 
144 145 146 147 148 149 150 151 152 153 154 
155 156 157 158 159 160 161 162 163 164 165 
166 167 168 169 170 171 172 173 174 175 176 
177 178 179 180 181 182 183 184 185 186 187 
188 189 190 191 192 193 194 195 196 197 198 
199 200 

# 有哪些知识点？

## 基础的数据结构题目

### HashSet and HashMap

### Stack, Queue, and Something Else

### LinkedList and BinaryTree and Tree

## 基础的算法题目

### 排序

### 回溯

## 基础的其他题目

### 位运算

## 进阶的算法题目

## 1 - Two-sum 家族 

Two-sum 如果考的话，最喜欢考的是 3sum 和 K-sum, 其他的考那就是白送分。

只需要重复 
015 和 560 就行了。         

典型题目：    
001，167,  015， 560.            
其中 167 是 015 的子问题，560是001的变种。     
Two-Sum 有两个阶段的问题，一个是需要解决无序数组 O(N)，此时必须要 O(N) 的 space.

* [001: Two-Sum](https://leetcode.com/problems/two-sum/)   
    Two Sum 最优解法不能Sort, HashMap 一遍过       
* [167: Two-Sum-II---Input-array-is-sorted](https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/)    
* [015: 3Sum](https://leetcode.com/problems/3sum/)   
    3 Sum 和 4 Sum 最优要Sort，为了避免重复的组合，连续重复的数字要取第一个，后面重复的是可以出现在组合中的。
* [560: Subarray-Sum-Equals-K](https://leetcode.com/problems/subarray-sum-equals-k/)      
* [523: Continuous-Subarray-Sum](https://leetcode.com/problems/continuous-subarray-sum/)

* [018: 4Sum](https://leetcode.com/problems/4sum/) 
4 Sum 是不一样的！因为 4 Sum 的解，解决了 k sum 的问题！
https://blog.csdn.net/bhwolf1987/article/details/14104589 
 
**一个问题是 如何解释 K-sum**    
Suppose we already know 2 sum. If k == 2, we just return the 2 sum.  
Now we firstly consider 3 sum. Consider the index i, denote the subarray starts from i + 1 as A_sub.   
The problems of 3 sum has transformed in to nums[i] + 2 sum solution set of A_sub with target of - nums[i].   
And we generate this idea to K-sum.  
K sum -> nums[i] + solution set of K - 1 sum with target of -nums[i].     

以下是一些没啥新意的：   
[259: 3Sum-Smaller](https://leetcode.com/problems/3sum-smaller/)   
[016: 3Sum-Closest](https://leetcode.com/problems/3sum-closest/)  
           
[653: Two-Sum-IV---Input-is-a-BST](https://leetcode.com/problems/two-sum-iv-input-is-a-bst/)     
[170: Two-Sum-III---Data-structure-design](https://leetcode.com/problems/two-sum-iii-data-structure-design/)   

## 2 - 处理运算进位的
08/08/2020     
445 和 415 是更典型的。 067 是 415 的 子集。   
445 用到 stack 还是很机智的。 可以做做。
* [002: Add-Two-Numbers](https://leetcode.com/problems/add-two-numbers/)      
* [067: Add-Binary](https://leetcode.com/problems/add-binary/)
* [415: Add-Strings](https://leetcode.com/problems/add-strings/)
* [445: Add-Two-Numbers-II](https://leetcode.com/problems/add-two-numbers-ii/)

989 有点不一样，还没做，可以看看。                   
[989: Add-to-Array-Form-of-Integer](https://leetcode.com/problems/add-to-array-form-of-integer/)

## 3 - Sliding Window   
08/08/2020   
// 以下是一组，越来越难，其中003由于最简单，所以可以把HashMap 简化为HashSet，而 159 没有做的意义。    
但是注意 003 的优化可以用 HashMap 直接记录每个字符出现的最后一次的位置。     
340 当然也可以这么做，但是就需要一个数组或者两个HashMap了。     
[003: Longest-Substring-Without-Repeating-Characters](https://leetcode.com/problems/longest-substring-without-repeating-characters/)    
[159: Longest-Substring-with-At-Most-Two-Distinct-Characters](https://leetcode.com/problems/longest-substring-with-at-most-two-distinct-characters/)       
[340: Longest-Substring-with-At-Most-K-Distinct-Characters](https://leetcode.com/problems/longest-substring-with-at-most-k-distinct-characters/)    
注意这一个题写 sliding window 的模板。
定义左指针，右指针直接循环就完事儿了，     

甭管行不行先加上，然后再倒退。   

以 340 为例： 
注意的是，这里可以进一步优化。
```java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        int size = 0;
        int max = 0;
        Map<Character, Integer> counts = new HashMap<>();
        
        char[] chars = s.toCharArray();
        int left = 0;
        
        for(int right = 0; right < s.length(); right++ ) {
            char c = chars[right];
            counts.put(c, counts.getOrDefault(c, 0) + 1);
            size += 1;
            while(counts.size() > k) {
                size -= 1;
                char last = chars[left++];
                counts.put(last, counts.get(last) - 1);
                if(counts.get(last) == 0) {
                    counts.remove(last);
                } 
            }
            max = Math.max(max, size);
        }
        return max;
    }
}
```

## 4 - Median-of-Two-Sorted-Arrays
目前只发现它自己, 真的挺难，值得背下来。           
利用了数组总长度的信息和有序的信息。

```
 1 3 | 4  5   
   l1  r1     
 1 2 | 5  7  8     
   l2  r2     
```
* [004: Median-of-Two-Sorted-Arrays](https://leetcode.com/problems/median-of-two-sorted-arrays/)      
笔记: [004. Median-of-Two-Sorted-Arrays](leetCode-004-Median-of-Two-Sorted-Arrays.md)

## 5 - Longest-Palindromic-Substring
写了，超级高频。写出从每个位置开始扩张的最长substring的长度，记住 i， i 和 i, i+ 1两种。然后记录最大的长度的中心位置和长度，
这样能倒推出应当substring的起始点。
* [005: Longest-Palindromic-Substring](https://leetcode.com/problems/longest-palindromic-substring/)
* [647: Palindromic-Substrings](https://leetcode.com/problems/palindromic-substrings/)

## 6 ZigZag-Conversion 
没做
[006: ZigZag-Conversion](https://leetcode.com/problems/zigzag-conversion/)

## 7 - Reverse-Integer 
虽然不难，值得一做。 007 要注意处理 越界的事儿。190要注意位运算有几个骚操作。
* [007: Reverse-Integer](https://leetcode.com/problems/reverse-integer/)
* [009: Palindrome-Number](https://leetcode.com/problems/palindrome-number/)      
    009 用 palindrome string 就没啥难度秒杀。 所以归类为 reverse integer.
* [190: Reverse-Bits](https://leetcode.com/problems/reverse-bits/)

## 8 String-to-Integer-(atoi)
* [008: String-to-Integer-(atoi)](https://leetcode.com/problems/string-to-integer-atoi/)

## 9 - 两个字符串对比的pattern
* [010: Regular-Expression-Matching](https://leetcode.com/problems/regular-expression-matching/) 高频
* [044: Wildcard-Matching](https://leetcode.com/problems/wildcard-matching/) 中频   
* [072: Edit-Distance](https://leetcode.com/problems/edit-distance/) 中频   

一个参考
https://www.cs.umd.edu/users/meesh/cmsc351/mount/lectures/lect25-longest-common-subseq.pdf    
* [115: Distinct-Subsequences](https://leetcode.com/problems/distinct-subsequences/)    低频         
* [1143: Longest-Common-Subsequence](https://leetcode.com/problems/longest-common-subsequence/) 中低频
* [1312: Minimum-Insertion-Steps-to-Make-a-String-Palindrome](https://leetcode.com/problems/minimum-insertion-steps-to-make-a-string-palindrome/)      低频
1312 和 1143 是几乎一样的。
* [516: Longest-Palindromic-Subsequence](https://leetcode.com/problems/longest-palindromic-subsequence/)      
    516 和 1312 是正儿八经的一模一样。   
    
还有还没归类的 
* [1092: Shortest-Common-Supersequence](https://leetcode.com/problems/shortest-common-supersequence/) 低频

## 10 - Container-With-Most-Water 高频
* [011: Container-With-Most-Water](https://leetcode.com/problems/container-with-most-water/)    
与 042 有那么一点相似吧，不合并了。     
重点是一句 短板决定水量，移动短板。

## 11 Integer <=> Roman
* [012: Integer-to-Roman](https://leetcode.com/problems/integer-to-roman/)
* [013: Roman-to-Integer](https://leetcode.com/problems/roman-to-integer/)

## Permutations/Combinations/Subsets
基础的 BackTrack 套件   
Permutations: TO(N!);       
Combinations: TO( N!/(N-k)!/k! );     
Subsets: TO(N*2^N)       

46 高频！  
* [046: Permutations](https://leetcode.com/problems/permutations/)
* [047: Permutations-II](https://leetcode.com/problems/permutations-ii/)  有重复数字
* [077: Combinations](https://leetcode.com/problems/combinations/)   
* [078: Subsets](https://leetcode.com/problems/subsets/) 实际上
* [090: Subsets-II](https://leetcode.com/problems/subsets-ii/)   有重复数字

17高频应当经常做做，尤其是初始化。   254 低频，但是应该是最难的。     
   
39 40 77 216 选一个做就好了。       
* [017: Letter-Combinations-of-a-Phone-Number](https://leetcode.com/problems/letter-combinations-of-a-phone-number/)
* [022: Generate-Parentheses](https://leetcode.com/problems/generate-parentheses/)
* [039: Combination-Sum](https://leetcode.com/problems/combination-sum/)
* [040: Combination-Sum-II](https://leetcode.com/problems/combination-sum-ii/)  

* [216: Combination-Sum-III](https://leetcode.com/problems/combination-sum-iii/)  
* [254: Factor-Combinations](https://leetcode.com/problems/factor-combinations/)      
     
注意 377 可以 combinations, 但是会 TLE, 所以是个 dp 了      
* [377: Combination-Sum-IV](https://leetcode.com/problems/combination-sum-iv/)




##  Merge-k-Sorted-Lists
23 里包括了 21    
* [021: Merge-Two-Sorted-Lists](https://leetcode.com/problems/merge-two-sorted-lists/)
* [023: Merge-k-Sorted-Lists](https://leetcode.com/problems/merge-k-sorted-lists/) 
* [088: Merge-Sorted-Array](https://leetcode.com/problems/merge-sorted-array/)   
    从尾部开始 merge 比较机智
* [977: Squares-of-a-Sorted-Array](https://leetcode.com/problems/squares-of-a-sorted-array/)     
    找到负的 subarray 的结尾和非负 subarray 的起点，然后直接就能 merge 一个 O(N) 的结果。

## Next Permutation 超高频 
暂时无相关的题。
这个题重在分析题意, next permutation 是按照 lexicographically
字典顺序的下一个。
其实很难讲清楚。随便写点吧，不行就去参考解答。
54321 是 lexicographically 到头了，所以直接翻转。
如果 .....46521 是这么一个结尾，6521 是完全倒序的，是最大的字典顺序，
如果 4 这里是 大于等于 6 的，我们仍然是最大字典序，没什么好说的，但是这里
突然出现了一个 4，也就是说一个山顶出现了，那么我们是可以做 next permutation 的。
相当于我们要进位。那么我们找 6521 里面 smallest element that is greater than 4.
拿它和 4 交换 得到.....56421. 由于是 next permutation, 交换之后，由于 6421
仍然有序并且是最大的字典序，所以我们要翻转 6421.
* [031: Next-Permutation](https://leetcode.com/problems/next-permutation/)

## Search-in-Rotated-Sorted-Array   
081 确实更一般情况，但是 033 考的高频的多。核心的判断就是，确定左侧是否有序，然后直接 else 就行了。
153 是有一个不错的题。处理 pivot 很容易，但是如何返回呢？ 答案是 看 nums[pivot] < nums[pivot + 1] return pivot + 1, 
num[pivot] < nums[pivot - 1] return pivot. 等于一次检查前后一共三个点。     
154 是超级低频的。有Duplicate 的 binary search 都很难.
* [033: Search-in-Rotated-Sorted-Array](https://leetcode.com/problems/search-in-rotated-sorted-array/)
* [153: Find-Minimum-in-Rotated-Sorted-Array](https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/)


* [081: Search-in-Rotated-Sorted-Array-II](https://leetcode.com/problems/search-in-rotated-sorted-array-ii/)
* [154: Find-Minimum-in-Rotated-Sorted-Array-II](https://leetcode.com/problems/find-minimum-in-rotated-sorted-array-ii/)

## Count-and-Say
 这个题麻烦在于弄懂题意。看起来好像可以递归，实际上不太能。
 其实就是重复 n - 1 次 **把一个函数的输出当成下一次的输入**。
 而这个函数本身干了什么事，虽然是这个题目中比较重要的部分，但是却是这个题中得到的最不重要的知识。
 我觉得这个题目考察的最重要的就是，**能写出去耦的代码**。

* [038: Count-and-Say](https://leetcode.com/problems/count-and-say/)

## Multiply-Strings
* [043: Multiply-Strings](https://leetcode.com/problems/multiply-strings/)               
043 String 乘法 M * N, 123 * 456 
最重要的观察是 M 数的 i 位 和 N 数的 j 位相乘，得到的是 i + j 和 i + j + 1 位。

## Jump-Game
045 真是一道魔鬼题，可以用O(N^2)的DP方法解。 最优解竟然是一个类似于贪心的BFS。
Jump Game 其实挺死亡的，它本身是个DAG，所以用DFS的 backtracking 是可以的。   
但是时间复杂度非常高。
* [045: Jump-Game-II](https://leetcode.com/problems/jump-game-ii/)   
* [055: Jump-Game](https://leetcode.com/problems/jump-game/)   

## Anagrams
49 自然又高频又好啊，而438是一个sliding window, 当然我们还是给归到这里了。
* [049: Group-Anagrams](https://leetcode.com/problems/group-anagrams/)    
    49 一个亮点是 char[] 可以直接 String.valueOf 得到一个拼接 String, int[] 不行，int[] 是一个地址
* [242: Valid-Anagram](https://leetcode.com/problems/valid-anagram/)
* [438: Find-All-Anagrams-in-a-String](https://leetcode.com/problems/find-all-anagrams-in-a-string/)


## 处理时间间隔
763 和56很像 很机智。
08/05/2020      
* [056: Merge-Intervals](https://leetcode.com/problems/merge-intervals/)
* [057: Insert-Interval](https://leetcode.com/problems/insert-interval/) 
* [252: Meeting-Rooms](https://leetcode.com/problems/meeting-rooms/)   
    252 不重要
* [253: Meeting-Rooms-II](https://leetcode.com/problems/meeting-rooms-ii/)  
    253 高频，而且用PQ 确实很机智，很重要。   
* [452: Minimum-Number-of-Arrows-to-Burst-Balloons](https://leetcode.com/problems/minimum-number-of-arrows-to-burst-balloons/)   
    452 不高频，还挺有意思的，是056 反其道而行之，缩小区间而非连接区间。 
* [763: Partition-Labels](https://leetcode.com/problems/partition-labels/)
    763 高频，是merge interval的解法，同时有一个贪心的算法，也很不错。
* [729: My-Calendar-I](https://leetcode.com/problems/my-calendar-i/)  低频。   
 可以用 TreeMap, 然后 floorKey() 和 ceilingKey()。 TreeMap 的 解法比 List 时间复杂度好。    
 为了能 hold 住下一题，还要考虑用 List<int[]> 是一个 N^2。                 
* [731: My-Calendar-II](https://leetcode.com/problems/my-calendar-ii/)    低频           
 这一题就想着 252，搞了好几个 thread。按这个题的要求，那我们就2个thread 就好。
 一个存正常的部分，和729题一样，一个 overlap， 只存overlap. 这又像 452 了。     
[435. Non-overlapping-Intervals](leetCode-435-Non-overlapping-Intervals.md)   
  这个题不能用 253 的写法，其实是一个贪心Greedy。它和452的联系其实更紧密。 
  它贪心的地方在于，我们排序好之后，两个有重复部分的，我们永远只扔掉那个触角伸的更远的，因为它更大可能造成重复。 

## Rotate-List
* [061: Rotate-List](https://leetcode.com/problems/rotate-list/)    
* [189: Rotate-Array](https://leetcode.com/problems/rotate-array/)   
* [186: Reverse-Words-in-a-String-II](https://leetcode.com/problems/reverse-words-in-a-string-ii/)      
186 确实和 189 还有点像的

## 一些 2D 动态规划 高频
这些都挺高频都挺好的.      
174 特别好，是非常好的题！只能从终点往起点 dp, 和 minimum-path-sum 不一样         
062, 063,064 好理解，只查找它自己的上左邻居。221是查找自己的 上，左，西北邻居。
062-221 都是高频。
* [062: Unique-Paths](https://leetcode.com/problems/unique-paths/)    
* [063: Unique-Paths-II](https://leetcode.com/problems/unique-paths-ii/)  
* [064: Minimum-Path-Sum](https://leetcode.com/problems/minimum-path-sum/)
* [174: Dungeon-Game](https://leetcode.com/problems/dungeon-game/)     
* [221: Maximal-Square](https://leetcode.com/problems/maximal-square/)


他们和第10题那一坨很像。但是这个更简单。

## 动态规划 Fibonacci Like 1-D Dp
* [070: Climbing-Stairs](https://leetcode.com/problems/climbing-stairs/)  
* [091: Decode-Ways](https://leetcode.com/problems/decode-ways/)
* [509: Fibonacci-Number](https://leetcode.com/problems/fibonacci-number/)  

## Search a 2D matrix
* [074: Search-a-2D-Matrix](https://leetcode.com/problems/search-a-2d-matrix/)    
* [240: Search-a-2D-Matrix-II](https://leetcode.com/problems/search-a-2d-matrix-ii/)
    240用binary search 不是最佳，但是和074放一起比较好。

## Quick Sort 相关，部分可以 HeapSort
75, 215, 973, 347, 694 都很高频         
075 是一个用 Quick Sort 的方法， 215 最佳方案是一个类似于 Quick Sort 的 Quick Select。         
Quick Select 就是为 215 而生的。            
* [075: Sort-Colors](https://leetcode.com/problems/sort-colors/)  三向快排   
    写075的重点是，075是3-way快排, i 的左侧是小于等于分割值的元素，left 左侧是小于的元素，left是等于分割元素或者等于i的。     
    而right 右侧是大于分割的元素，right本身不确定。
    所以把左侧元素和i交换之后，可以 i++,left++,但是right和i交换之后，只有right--而不i++.
* [215: Kth-Largest-Element-in-an-Array](https://leetcode.com/problems/kth-largest-element-in-an-array/)     
215 笔记 
[215. Kth-Largest-Element-in-an-Array](leetCode-215-Kth-Largest-Element-in-an-Array.md)  

[中文力扣变形，找前k小的数](https://leetcode-cn.com/problems/zui-xiao-de-kge-shu-lcof/)
这个题和973差不多，由于不需要排序好的结果，用quick select找到第k大的时候直接就可以返回pivot前面的数组。

* [973: K-Closest-Points-to-Origin](https://leetcode.com/problems/k-closest-points-to-origin/)       
973 是 215 的换汤不换药，要保证能够起码做出 heap的解法，然后挑战 quick select.     
* [414: Third-Maximum-Number](https://leetcode.com/problems/third-maximum-number/) 
    414 是 215 的子集，没有什么意思。 但是它强迫你必须O(N)解法。而且第三大是不重复的第三大，因此比较有意思了。
* [703: Kth-Largest-Element-in-a-Stream](https://leetcode.com/problems/kth-largest-element-in-a-stream/)
    703 是 215 的另一类子集，最好的办法就是用 heap sort 做，这下不能用 quick sort 了。
* [692: Top-K-Frequent-Words](https://leetcode.com/problems/top-k-frequent-words/)      
* [347: Top-K-Frequent-Elements](https://leetcode.com/problems/top-k-frequent-elements/)     
    347 和 692 其实一样, 原则上都可以 quick select 频率，但是347题目保证了不会有 第 k 个有好几个的情况，此时692就很恶心了。因为692还要考虑如果第 k 个有好几个，输出哪一个的问题。
    347 因此还可以用 bucket sort 很轻松的写。
    692 因此很高频。所以至少要保证692可以做出 heap 的解法。

## Minimum-Window-Substring 高频，滑窗
* [076: Minimum-Window-Substring](https://leetcode.com/problems/minimum-window-substring/)

## Word Search I & II
* [079: Word-Search](https://leetcode.com/problems/word-search/) 高频
* [212: Word-Search-II](https://leetcode.com/problems/word-search-ii/)  中频           
其实这已经是两个题了。word-search I 就是一个普通的 dfs 回溯，而 word-search-ii 是个结合了 trie 的 dfs 回溯。
当然这两个题都是很好的题，虽然有其他很相似的题，我们也不合并到其他地方了。

## Gray Code
这个题的难点在于如何精巧的实现每次只变一个数
* [089: Gray-Code](https://leetcode.com/problems/gray-code/) 

## Palindrome
[125: Valid-Palindrome](https://leetcode.com/problems/valid-palindrome/)     
[680: Valid-Palindrome-II](https://leetcode.com/problems/valid-palindrome-ii/)     

## Word-Ladder 高频
* [127: Word-Ladder](https://leetcode.com/problems/word-ladder/)
* [126: Word-Ladder-II](https://leetcode.com/problems/word-ladder-ii/)

## 纯考 Bitwise 的系列题目 Single Numbers
这三个题太好了！刷熟练。

代码笔记: 
* [136. Single-Number](leetCode-136-Single-Number.md)
* [137. Single-Number-II](leetCode-137-Single-Number-II.md)
* [260. Single-Number-III](leetCode-260-Single-Number-III.md)

题目：
* [136: Single-Number](https://leetcode.com/problems/single-number/)  
* [137: Single-Number-II](https://leetcode.com/problems/single-number-ii/)   
* [260: Single-Number-III](https://leetcode.com/problems/single-number-iii/)     

137, 260 用HashMap或者是排序的方法都很平庸，必须要能意识到用 bitwise operation 才能过关。

## Word Break 高频 是一个 BFS ？
* [139: Word-Break](https://leetcode.com/problems/word-break/)    
用 BFS 把 s 分解为 子数组解决。例如 从 0 到 4 匹配了一个词，那么就从 4 开始重复这个问题。
一旦 4 开始的 子问题解决了，我们就 mark 一下。这样避免重复。
* [140: Word-Break-II](https://leetcode.com/problems/word-break-ii/)    不会

## Floyd Find Cycle, 快慢指针
* [141: Linked-List-Cycle](https://leetcode.com/problems/linked-list-cycle/)     
* [142: Linked-List-Cycle-II](https://leetcode.com/problems/linked-list-cycle-ii/)      
* [160: Intersection-of-Two-Linked-Lists](https://leetcode.com/problems/intersection-of-two-linked-lists/)      
* [287: Find-the-Duplicate-Number](https://leetcode.com/problems/find-the-duplicate-number/)  

## Design: LRU-cache 高频
* [146: LRU-Cache](https://leetcode.com/problems/lru-cache/) 超高频   
* [460: LFU-Cache](https://leetcode.com/problems/lfu-cache/) 中高频

## Design: Min-Stack 高频
* [155: Min-Stack](https://leetcode.com/problems/min-stack/)  高频
* [716: Max-Stack](https://leetcode.com/problems/max-stack/)  中频

## Find-Peak-Element
* [162: Find-Peak-Element](https://leetcode.com/problems/find-peak-element/)    
* [852: Peak-Index-in-a-Mountain-Array](https://leetcode.com/problems/peak-index-in-a-mountain-array/)     

## Majority-Element   
Majority 的常规做法就是 HashMap. 但是 169， 229 还有 一个 voting algorithm。 学学。
1150 因为有序，所以可以搞个 log(N) 的法子。       
* [169: Majority-Element](https://leetcode.com/problems/majority-element/)     
* [229: Majority-Element-II](https://leetcode.com/problems/majority-element-ii/)     
* [1150: Check-If-a-Number-Is-Majority-Element-in-a-Sorted-Array](https://leetcode.com/problems/check-if-a-number-is-majority-element-in-a-sorted-array/)

## 位运算
* [191: Number-of-1-Bits](https://leetcode.com/problems/number-of-1-bits/)

## House Robber 
213 比 198 更普遍适用。                 
* [198: House-Robber](https://leetcode.com/problems/house-robber/)
* [213: House-Robber-II](https://leetcode.com/problems/house-robber-ii/)


256 和 265 的 dp 解法挺好的。好好想想，确实和 House Robber 有一定的相似之处。        
* [256: Paint-House](https://leetcode.com/problems/paint-house/)
* [265: Paint-House-II](https://leetcode.com/problems/paint-house-ii/)          


说实话337是个很好的树的问题，但是不太好归类。就留给这一题了。       
* [337: House-Robber-III](https://leetcode.com/problems/house-robber-iii/)

## 找联通分量个数 number of connected pieces, DFS, BFS, UnionFind 皆宜
* [200: Number-of-Islands](https://leetcode.com/problems/number-of-islands/)
* [305: Number-of-Islands-II](https://leetcode.com/problems/number-of-islands-ii/)     
    305是一个UnionFind 最佳的题              
* [547. Friend-Circles](leetCode-547-Friend-Circles.md)

## Reverse LinkedList
都是不错的题目。 206 和 156 更接近。
* [092: Reverse-Linked-List-II](https://leetcode.com/problems/reverse-linked-list-ii/)
* [206: Reverse-Linked-List](https://leetcode.com/problems/reverse-linked-list/)    
* [234: Palindrome-Linked-List](https://leetcode.com/problems/palindrome-linked-list/)  
* [156: Binary-Tree-Upside-Down](https://leetcode.com/problems/binary-tree-upside-down/)    




## Trie
* [208: Implement-Trie-(Prefix-Tree)](https://leetcode.com/problems/implement-trie-prefix-tree/)
* [642: Design-Search-Autocomplete-System](https://leetcode.com/problems/design-search-autocomplete-system/)   
* [211: Design-Add-and-Search-Words-Data-Structure](https://leetcode.com/problems/design-add-and-search-words-data-structure/)  


## Two Pointer 类似 Monotonic Queue/ Array 的

[209: Minimum-Size-Subarray-Sum](https://leetcode.com/problems/minimum-size-subarray-sum/)    


## 拓扑排序   
* [207: Course-Schedule](https://leetcode.com/problems/course-schedule/)          
* [210: Course-Schedule-II](https://leetcode.com/problems/course-schedule-ii/)      
* [261: Graph-Valid-Tree](https://leetcode.com/problems/graph-valid-tree/)       
    为什么把261也放进来呢？因为树就是一个DAG，要判断有无环，你环都判断了，拓扑排序就水到渠成了。
* [310: Minimum-Height-Trees](https://leetcode.com/problems/minimum-height-trees/)          
    为什么把 310 放进来呢？因为310的 degrees bfs法和 bfs的拓扑排序很像

## Basic Calculator    高频
典型的是772.    
227 没有括号，更简单。224, 227 因为有括号，所以可以搞个 dfs.   
处理 String 的方法都是 sum + preNum preOp num c.      
根据 c 的不同，决定怎么处理 sum + preNum preOp num   并更新指针。         
易错点有要加个 dummyEnd。      
* [224: Basic-Calculator](https://leetcode.com/problems/basic-calculator/)    
* [227: Basic-Calculator-II](https://leetcode.com/problems/basic-calculator-ii/)     
* [772: Basic-Calculator-III](https://leetcode.com/problems/basic-calculator-iii/)    
Notes:   [772. Basic-Calculator-III](leetCode-772-Basic-Calculator-III.md)     

## Lowest Common Ancestor   
KeyWord: PostOrder     
236 比较重要和高频，1123 更难一点。基本的想法都是 postOrder 之后再处理。
235由于是 BST可以利用 BST的性质做到 O(1) 的 space complexity.  
* [235: Lowest-Common-Ancestor-of-a-Binary-Search-Tree](https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/)   
* [236: Lowest-Common-Ancestor-of-a-Binary-Tree](https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/)
* [1123: Lowest-Common-Ancestor-of-Deepest-Leaves](https://leetcode.com/problems/lowest-common-ancestor-of-deepest-leaves/)

## 239: Sliding-Window-Maximum 
Sliding Window 并且 monotonic queue.         
* [239: Sliding-Window-Maximum](https://leetcode.com/problems/sliding-window-maximum/)

## 263 Ugly Number 丑数 

264 是一个挺有意思的动态规划的题目，至少要保证想出来heap的解法。
* [263: Ugly-Number](https://leetcode.com/problems/ugly-number/)    
* [264: Ugly-Number-II](https://leetcode.com/problems/ugly-number-ii/)    
* [313: Super-Ugly-Number](https://leetcode.com/problems/super-ugly-number/)

如果参透了 264，那么313， 都不难

## 一个单独的 1D dp
* [279: Perfect-Squares](https://leetcode.com/problems/perfect-squares/)

## LIS: 高频并且重点。
* [300: Longest-Increasing-Subsequence](https://leetcode.com/problems/longest-increasing-subsequence/)
* [368: Largest-Divisible-Subset](https://leetcode.com/problems/largest-divisible-subset/)

673 有 Segment Tree 的解法。
* [673: Number-of-Longest-Increasing-Subsequence](https://leetcode.com/problems/number-of-longest-increasing-subsequence/)

随手做了一个 easy 水题
* [674: Longest-Continuous-Increasing-Subsequence](https://leetcode.com/problems/longest-continuous-increasing-subsequence/)

## Sparse Matrix   中频       
* [311: Sparse-Matrix-Multiplication](https://leetcode.com/problems/sparse-matrix-multiplication/)         
矩阵乘法:     
左边 m1*n1 的矩阵 和 m2 * n2 相乘，有 n1 == m2. 结果矩阵是 m1*n2的         
左边矩阵 (i1, j1) 元素和, 右边矩阵 (i2, j2) 的积，是加在结果矩阵的 （i1, j2) 位置上的，注意 j1 == i2 。    
所以 左侧 (i1, j1) 是找 右侧 j1 行所有的元素 (j1, j2) 相乘并且加在 (i1, j2) 上面。            
因此我们可以把 左右矩阵 用 HashMap 存起来。    
其中左矩阵 integer 是行，按行存元素。   
右矩阵 integer 是列，按列存元素。   
这样我们从左矩阵取出一个数(i1,j1)的时候，直接就找到右矩阵所有的 j1 列，      
然后把结果加到 (i1, j2) 上面。      
```java
Map<Integer, List<Integer, Integer > > leftMatrixRows;
Map<Integer, List<Integer, Integer > > rightMatrixCols;
Map<Integer, Map<Integer, Integer>> ans;
```
参考稀疏矩阵的表示    
https://www.geeksforgeeks.org/sparse-matrix-representations-using-list-lists-dictionary-keys/?ref=rp   

* [1570: Dot-Product-of-Two-Sparse-Vectors](https://leetcode.com/problems/dot-product-of-two-sparse-vectors/)    
稀疏向量

## Design: Insert-Delete-GetRandom-O(1) 高频
核心，HashMap + ArrayList
* [380: Insert-Delete-GetRandom-O(1)](https://leetcode.com/problems/insert-delete-getrandom-o1/)     高频      
* [381: Insert-Delete-GetRandom-O(1)---Duplicates-allowed](https://leetcode.com/problems/insert-delete-getrandom-o1-duplicates-allowed/)

### 讨厌啊处理字符串，数字的 没搞明白其实
* [402: Remove-K-Digits](https://leetcode.com/problems/remove-k-digits/)   

## Design File System: 真的还挺难的 高频
* [588: Design-In-Memory-File-System](https://leetcode.com/problems/design-in-memory-file-system/)     
 核心是建立一个File 节点。最好可以区分自己是 file 还是 folder.   

## Stack 相关
[735: Asteroid-Collision](https://leetcode.com/problems/asteroid-collision/)    

## Reorganize-String 高频
这个题 O(N) 很好玩。和621 有点像的。
* [767: Reorganize-String](https://leetcode.com/problems/reorganize-string/)
* [621: Task-Scheduler](https://leetcode.com/problems/task-scheduler/)   中高频

## 左右括号相关
[1249: Minimum-Remove-to-Make-Valid-Parentheses](https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/)